src/components/TodoList.tsx

```diff
@@ -0,0 +1,204 @@
+import {
+  ChangeEvent, FocusEvent, FormEvent,
+  KeyboardEvent, useEffect, useRef, useState,
+} from 'react';
+import classNames from 'classnames';
+import { deleteTodo, updateTodo } from '../api/todos';
+import { Filter } from '../types/Filter';
+import { Todo } from '../types/Todo';
+
+interface Props {
+  todos: Todo[],
+  filter: Filter,
+  tempTodo: Todo | null,
+  handleError: (error: string) => void,
+  isUpdating: boolean,
+  handleIsUpdating: (status: boolean) => void,
+  updatingIds: number[],
+  handleUpdatingIds: (ids: number[]) => void,
+  handleLoadTodos: () => void,
+}
+
+export const TodoList: React.FC<Props> = ({
+  todos, filter, tempTodo, handleError, handleLoadTodos,
+  isUpdating, handleIsUpdating, updatingIds, handleUpdatingIds,
+}) => {
+  const [editTodo, setEditTodo] = useState<Todo | null>(null);
+  const [newTitle, setNewTitle] = useState('');
+
+  useEffect(() => {
+    handleLoadTodos();
+  }, [isUpdating]);
+
+  const titleRef = useRef<HTMLInputElement | null>(null);
+
+  const filteredTodos = todos.filter(todo => {
+    switch (filter) {
+      case Filter.All:
+        return true;
+      case Filter.Active:
+        return !todo.completed;
+      case Filter.Completed:
+        return todo.completed;
+      default:
+        throw new Error('wrong filter selected');
+    }
+  });
+
+  const handleCleaner = () => {
+    handleIsUpdating(false);
+    handleUpdatingIds([]);
+  };
+
+  const handleDelete = (id: number) => {
+    handleIsUpdating(true);
+    handleUpdatingIds([id]);
+    deleteTodo(id)
+      .then(() => handleCleaner())
+      .catch(() => handleError('Unable to delete a todo'));
+  };
+
+  const handleUpdateCompleted = (id:number, completed: boolean) => {
+    handleIsUpdating(true);
+    handleUpdatingIds([id]);
+    const updatedTodo = {
+      completed: !completed,
+    };
+
+    updateTodo(id, updatedTodo)
+      .then(() => handleCleaner())
+      .catch(() => handleError('Unable to update a todo'));
+  };
+
+  const handleEditToto = (todo: Todo) => {
+    setEditTodo(todo);
+    setNewTitle(todo.title);
+  };
+
+  const handleTitleSubmit = (
+    event: FormEvent<HTMLFormElement> | FocusEvent<HTMLInputElement, Element>,
+    id: number,
+  ) => {
+    event.preventDefault();
+    if (!newTitle.length || !editTodo) {
+      handleDelete(id);
+
+      return;
+    }
+
+    const editId = editTodo.id;
+    const newTodo = {
+      title: newTitle,
+    };
+
+    handleIsUpdating(true);
+    handleUpdatingIds([editId]);
+    setEditTodo(null);
+    updateTodo(editId, newTodo)
+      .then(handleCleaner)
+      .catch(() => handleError('Unable to update a todo'));
```

Comment:
Here you need to use `updateTodo` function from the app component with similar logic.